data_structure.py

```diff
@@ -237,6 +253,18 @@ def sv_zip(*iterables):
 #     ("REPEAT", "Repeat Last", "Match longest List by repeating last item",     3),
 #     ]
 
+def numpy_full_list(array, maxl):
+    # difl = maxl - array.shape[0]
+    try:
+        difl = maxl - array.shape[0]
+    except:
+        array = np.array(array)
+        difl = maxl - array.shape[0]
+    difl = maxl - len(array)
+    if difl > 0:
+        new_part = np.repeat(array[np.newaxis, -1], difl, axis=0)
+        array = np.concatenate((array, new_part))
+    return array[:maxl]
```

Comment:
Not too wide exception?
I would consider about possibility of using `np.resize` function here.
Instead of `np.repeat` using of `np.full` looks like more logical.

Reply:
np.resize has some complications with 2-axis arrays (vector arrays) also np.full would just work with 1d arrays.
About the try/ except y change it to `if type(array) != np.ndarray` that seems more readable :)

Reply:
Why don't use `AttributeError`?
It looks work:
```Python
np.full((3, 3), [1,2,3])
array([[1, 2, 3],
       [1, 2, 3],
       [1, 2, 3]])

a = np.array([[1,2,3], [1,2,3]])
a.resize(3,3)
a
array([[1, 2, 3],
       [1, 2, 3],
       [0, 0, 0]])
```

Reply:
```
np.full((3, 3), [[1,2,3],[2,3,4]])
ValueError: could not broadcast input array from shape (2,3) into shape (3,3)

```
This alone is awesome for cycle match
```
np.resize([[1,2,3],[2,3,4]], (5,3))
array([[1, 2, 3],
       [2, 3, 4],
       [1, 2, 3],
       [2, 3, 4],
       [1, 2, 3]]
       )
```


So i was testing this two
```
def numpy_full_list(array, maxl):
    if type(array) != np.ndarray:
        array = np.array(array)

    difl = maxl - array.shape[0]

    if difl > 0:
        new_part = np.repeat(array[np.newaxis, -1], difl, axis=0)
        array = np.concatenate((array, new_part))
    return array[:maxl]

def numpy_full_list2(array, maxl, vect=False):
    l= len(array)
    difl = maxl - l
    
    if difl > 0:
        if vect:
            array = np.resize(array, (maxl,3))
        else: 
            array = np.resize(array, maxl)
        array[l:]= array[l-1]
        
    return array[:maxl]
```
for 1d arrays the resize method is 2 or 3 times faster
for 2d arrays the concatenate method is  6 or 7 times faster
So i think i should keep the two of them and use them wisely
Any idea to improve them?

Reply:
There is no need in slicing:
```Python
np.resize([1,2,3,4], 2)
array([1, 2])
```
`Vector` argument is not good. Using of something like this should help.
```Python
np.resize(a, [length] + a.shape[1:])
```
The same is fair about `full` function:
```Python
a = np.array([[1,2],[3,4]])
np.full([5] + list(a.shape[1:]), a[-1])

array([[3, 4],
       [3, 4],
       [3, 4],
       [3, 4],
       [3, 4]])

a = np.array([[[1,2]], [[3,4]]])
np.full([5] + list(a.shape[1:]), a[-1])

array([[[3, 4]],
       [[3, 4]],
       [[3, 4]],
       [[3, 4]],
       [[3, 4]]])
```

Also can be used =)
```Python
np.pad(np.arange(5), (0,2), 'edge')
array([0, 1, 2, 3, 4, 4, 4])
```

Reply:
After some more test the repeat + concatenate method is in most of the cases faster than reshape and more flexible than full, so i think i'm going to leave it like that for the moment.
Anyway thanks for the feedback 👍